### ISSUE TITLE
Always explicitly specify function return types and property types when writing libraries

### ISSUE THREAD

**Issue by egor-n** [arturbosch/detekt#1404]
## Expected Behavior of the rule
Kotlin's official coding conventions tell us to explicitly specify function return type and property types to avoid accidentally exposing declarations as public API.

```kotlin
// Correct
fun show(): Unit {
  //
}

val myInt: Int = 1

private val myString = "test"
```

```kotlin
// Incorrect
fun show() {
  //
}

val myInt = 1
```

## Context
https://kotlinlang.org/docs/reference/coding-conventions.html#coding-conventions-for-libraries

**Comment by arturbosch** [comment]
Should be fairly easy to implement. However we can't configure specific paths to some rules.
If you have a mix of api and other modules, there need to be a filter somehow.
We could introduce a config property which accepts path regexes and tests `KtFile`'s in the `beforeVisit`-block.

**Comment by vanniktech** [comment]
Similiar to #462 ?

**Comment by arturbosch** [comment]
Closing and linking this one to #462 (thanks @vanniktech )
